model/service.go

```diff
@@ -21,40 +153,183 @@ import (
+
+// Endpoint defines a IP:port associated with an instance of the service. This is the address to which the
+// request/connection from the caller will be routed to. If a service has multiple ports, then the same
+// instance IP is listening on multiple ports (one corresponding to each service port). However,
+// we do not group such ports here. Instead we use a one endpoint per service port.
+
+// For e.g., if catalog.mystore.com is accessible through port 80 and 8080,
+// and it maps to an instance with IP 172.16.0.1, such that connections to
+// port 80 are forwarded to port 55446, and connections to port 8080 are
+// forwarded to port 33333,
+
+// then internally, we have two two endpoint structs for the
+// service catalog.mystore.com
+//  --> 172.16.0.1:54546 (with ServicePort pointing to 80) and
+//  --> 172.16.0.1:33333 (with ServicePort pointing to 8080)
+type Endpoint struct {
+	// Address of the network endpoint, typically an IPv4 address
+	Address string `json:"ip_address,omitempty"`
+
+	// Port number where this instance is listening for connections This
+	// need not be the same as the port where the service is accessed.
+	// e.g., catalog.mystore.com:8080 -> 172.16.0.1:55446
+	Port int `json:"port"`
+
+	// Port declaration from the service declaration This is the port for
+	// the service associated with this instance (e.g.,
+	// catalog.mystore.com)
+	ServicePort *Port `json:"port"`
+}
+
+// Tag is non empty set of key=value pair attributes (e.g., Labels in
+// kubernetes/mesos) that are assigned to pods/VMs etc. Each version of a
+// service can be uniquely identified by a set of tags
+type Tag map[string]string
+
+// TagList is a set of tags
+
+// FIXME rename to something else for clarity. Its not clear why one needs
+// TagList when Tag by itself is a set of key=value pairs
+type TagList []Tag
+
+// ServiceInstance represents an individual instance of a specific version
+// of a service. It binds a network endpoint (ip:port), the service
+// description (which is oblivious to various versions) and a set of tags
+// that describe the service version associated with this instance.
+
+// So, if catalog.mystore.com has two versions v1, v2, with instance IP
+// addresses 172.16.0.1:8888, 172.16.0.2:8888 of version v1 (tags foo=bar),
+
+// and instance IP addresses 172.16.0.3:8888, 172.16.0.4:8888 of version v2
+// (with tags kitty=kat),
+
+// the set of service instances for catalog.mystore.com are
+// catalog.myservice.com ->
+//      --> Endpoint(172.16.0.1:8888), Service(catalog.myservice.com), Tag(foo=bar)
+//      --> Endpoint(172.16.0.2:8888), Service(catalog.myservice.com), Tag(foo=bar)
+//      --> Endpoint(172.16.0.3:8888), Service(catalog.myservice.com), Tag(kitty=cat)
+//      --> Endpoint(172.16.0.4:8888), Service(catalog.myservice.com), Tag(kitty=cat)
+type ServiceInstance struct {
+	Endpoint Endpoint `json:"endpoint,omitempty"`
+	Service  *Service `json:"service,omitempty"`
+	Tag      Tag      `json:"tag,omitempty"`
```

Comment:
we need to rework this structure so not sure you want to document it now.
there is a missing concept of tag matching.
tags form a lattice by inclusion and that's not documented.